c++/src/kj/async-inl.h

```diff
@@ -1084,13 +1084,14 @@ inline PromiseForResult<Func, void> evalNow(Func&& func) {
 template <typename Func>
 struct RetryOnDisconnect_ {
   static inline PromiseForResult<Func, void> apply(Func&& func) {
-    auto promise = evalNow(func);
-    return promise.catch_([func = kj::mv(func)](kj::Exception&& e) -> PromiseForResult<Func, void> {
-      if (e.getType() == kj::Exception::Type::DISCONNECTED) {
-        return func();
-      } else {
-        return kj::mv(e);
-      }
+    return evalLater([func = kj::mv(func)]() mutable -> PromiseForResult<Func, void> {
+      return func().catch_([func = kj::mv(func)](kj::Exception&& e) mutable -> PromiseForResult<Func, void> {
```

Comment:
I think we still want `evalNow()` around `func()` here, so that any synchronously-thrown exceptions are caught too.

Reply:
Also note that the reason I used two separate lines before, like:

    auto promise = evalNow(func);
    return promise.catch_(...)

Rather than:

    return evalNow(func).catch_(...)

Is because, technically, in the one-line version, it's undefined whether lambda passed to `catch_` is evaluated before or after `evalNow(func)`. The lambda captures `func` by move, though, so if the lambda is evaluated first, then `func` is moved away and `evalNow(func)` will not work.

I have actually observed GCC using this unexpected ordering of evaluation in the past, though Clang seems to follow intuitive order.

Reply:
Yeah, good call. Added.

Reply:
Interesting, thanks. I had assumed that was well-defined. Changed back to separate lines.